src/coreclr/jit/lclvars.cpp

```diff
@@ -4181,49 +4181,62 @@ void Compiler::lvaMarkLclRefs(GenTree* tree, BasicBlock* block, Statement* stmt,
 
-            if (op1->gtOper == GT_LCL_VAR && op2->gtType != TYP_BOOL)
+            if (op1->gtOper == GT_LCL_VAR)
             {
-                /* Only simple assignments allowed for booleans */
+                LclVarDsc* varDsc = lvaGetDesc(op1->AsLclVarCommon());
 
-                if (tree->gtOper != GT_ASG)
+                if (varDsc->lvPinned && varDsc->lvAllDefsAreNoGc)
                 {
-                    goto NOT_BOOL;
+                    if (!op2->IsNotGcDef())
+                    {
+                        varDsc->lvAllDefsAreNoGc = false;
+                    }
                 }
 
-                /* Is the RHS clearly a boolean value? */
-
-                switch (op2->gtOper)
+                if (op2->gtType != TYP_BOOL)
                 {
-                    unsigned lclNum;
+                    /* Only simple assignments allowed for booleans */
 
-                    case GT_CNS_INT:
+                    if (tree->gtOper != GT_ASG)
+                    {
+                        goto NOT_BOOL;
+                    }
 
-                        if (op2->AsIntCon()->gtIconVal == 0)
-                        {
-                            break;
-                        }
-                        if (op2->AsIntCon()->gtIconVal == 1)
-                        {
-                            break;
-                        }
+                    /* Is the RHS clearly a boolean value? */
 
-                        // Not 0 or 1, fall through ....
-                        FALLTHROUGH;
+                    switch (op2->gtOper)
+                    {
+                        unsigned lclNum;
 
-                    default:
+                        case GT_CNS_INT:
 
-                        if (op2->OperIsCompare())
-                        {
-                            break;
-                        }
+                            if (op2->AsIntCon()->gtIconVal == 0)
+                            {
+                                break;
+                            }
+                            if (op2->AsIntCon()->gtIconVal == 1)
+                            {
+                                break;
+                            }
 
-                    NOT_BOOL:
+                            // Not 0 or 1, fall through ....
+                            FALLTHROUGH;
 
-                        lclNum = op1->AsLclVarCommon()->GetLclNum();
-                        noway_assert(lclNum < lvaCount);
+                        default:
 
-                        lvaTable[lclNum].lvIsBoolean = false;
-                        break;
+                            if (op2->OperIsCompare())
+                            {
+                                break;
+                            }
+
+                        NOT_BOOL:
+
+                            lclNum = op1->AsLclVarCommon()->GetLclNum();
+                            noway_assert(lclNum < lvaCount);
+
+                            lvaTable[lclNum].lvIsBoolean = false;
```

Comment:
Since you've fetched the varDsc above, you might as well use it here:
```suggestion
                            varDsc->lvIsBoolean = false;
```